packages/lisk-transactions/src/transactions/3_vote_transaction.ts

```diff
@@ -0,0 +1,423 @@
+			voteAssetFormatSchema,
+			this.asset,
+		) as boolean;
+		const errors = validator.errors
+			? validator.errors.map(
+					error =>
+						new TransactionError(
+							`'${error.dataPath}' ${error.message}`,
+							this.id,
+							error.dataPath,
+						),
+			  )
+			: [];
+
+		return {
+			id: this.id,
+			status:
+				status === Status.OK && valid && errors.length === 0
+					? Status.OK
+					: Status.FAIL,
+			errors,
+		};
+	}
+
+	public verify({
+		sender,
+		dependentState,
+	}: RequiredState): TransactionResponse {
+		const { errors: baseErrors } = super.apply({ sender });
+		if (!dependentState || !dependentState[ENTITY_ACCOUNT]) {
+			throw new Error('Dependent state is required for vote transaction');
+		}
+		const errors = [...baseErrors];
+		const dependentAccounts = dependentState[ENTITY_ACCOUNT];
+		if (!dependentAccounts) {
+			throw new Error('Entity account is required.');
+		}
+		if (!isAccountArray(dependentAccounts)) {
+			throw new Error('Required state does not have valid account type');
+		}
+		dependentAccounts.forEach(({ publicKey, username }) => {
+			if (
+				!this.asset.votes.map(vote => vote.substring(1)).includes(publicKey)
+			) {
+				errors.push(
+					new TransactionError(
+						`Dependent state for ${publicKey} is required.`,
+						this.id,
+					),
+				);
+			}
+
+			if (username === undefined || username === '') {
+				errors.push(
+					new TransactionError(`${publicKey} is not a delegate.`, this.id),
+				);
+			}
+		});
+		const senderVotes = sender.votes || [];
+		this.asset.votes.forEach(vote => {
+			const action = vote.charAt(0);
+			const publicKey = vote.substring(1);
+			// Check duplicate votes
+			if (action === PREFIX_UPVOTE && senderVotes.includes(publicKey)) {
+				errors.push(
+					new TransactionError(`${publicKey} is already voted.`, this.id),
+				);
+				// Check non-existing unvotes
+			} else if (action === PREFIX_UNVOTE && !senderVotes.includes(publicKey)) {
+				errors.push(
+					new TransactionError(`${publicKey} is not voted.`, this.id),
+				);
+			}
+			throw new Error('It should never reach here');
+		});
+
+		const upvotes = this.asset.votes.filter(
+			vote => vote.charAt(0) === PREFIX_UPVOTE,
+		);
+		const unvotes = this.asset.votes.filter(
+			vote => vote.charAt(0) === PREFIX_UNVOTE,
+		);
+		const votes: ReadonlyArray<string> = [...senderVotes, ...upvotes].filter(
+			vote => unvotes.includes(vote),
+		);
+		if (votes.length > MAX_VOTE_PER_ACCOUNT) {
+			errors.push(
+				new TransactionError(
+					`Vote cannot exceed ${MAX_VOTE_PER_ACCOUNT} but has ${votes.length}`,
+					this.id,
+				),
+			);
+		}
+
+		return {
+			id: this.id,
+			status: errors.length === 0 ? Status.OK : Status.FAIL,
+			errors,
+		};
+	}
+
+	public apply({ sender }: RequiredState): TransactionResponse {
+		const { errors: baseErrors, state } = super.apply({ sender });
+		if (!state) {
+			throw new Error('state is required for applying transaction');
```

Comment:
not TransactionError with a transaction id?

Reply:
TransactionError is handleable error (like it could return the message through API) but in this case, this error is more like a developer error, which should never happen. That's why i put it as standard error